core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1264,14 +1280,24 @@ private[spark] class DAGScheduler(
     listenerBus.post(SparkListenerTaskGettingResult(taskInfo))
   }
 
-  private[scheduler] def handleJobSubmitted(jobId: Int,
+  private[scheduler] def handleJobSubmitted(
+      jobId: Int,
       finalRDD: RDD[_],
       func: (TaskContext, Iterator[_]) => _,
       partitions: Array[Int],
       callSite: CallSite,
       listener: JobListener,
       artifacts: JobArtifactSet,
       properties: Properties): Unit = {
+    // If this job belongs to a cancelled job group, skip running it
+    val jobGroupIdOpt = Option(properties).map(_.getProperty(SparkContext.SPARK_JOB_GROUP_ID))
+    if (jobGroupIdOpt.exists(cancelledJobGroups.contains(_))) {
```

Comment:
Because the schedule thread and event thread are asynchronous, it seems doesn't cancel the future job not always effective.

Reply:
The `handler` function are run in the synchronized block, so the "`cancelJobGroup(cancelFutureJobs = true)` => the current active jobs will be cancelled + this job group is added to the set" should be run atomically. And later the future job belonging to this job group will be cancelled on submission.

Reply:
+1, I agree with @anchovYu's assessment:

- This `handleJobSubmitted` method and the `handleJobGroupCancelled` method both run in the single-threaded event thread.
- Consider both potential interleavings:
  - If `handleJobGroupCancelled` runs first then the group will be marked as cancelled and `handleJobSubmitted` will skip the job.
  - If `handleJobSubmitted` runs first then `handleJobGroupCancelled` will see a running job and cancel it.

Reply:
@JoshRosen Thank you for the description. I got it now.